src/expr/src/typecheck.rs

```diff
@@ -0,0 +1,777 @@
+                                    expected: base_col.clone(),
+                                    message: format!(
+                                        "couldn't compute union of column types in let rec: {:?}",
+                                        e
+                                    ),
+                                }
+                            })?;
+                        }
+                    } else {
+                        ctx.insert(id, typ);
+                    }
+                }
+
+                body.typecheck(&mut ctx)
+            }
+            ArrangeBy { input, keys } => {
+                let t_in = input.typecheck(ctx)?;
+
+                for cols in keys {
+                    for col in cols {
+                        let _ = col.typecheck(self, &t_in)?;
+                    }
+                }
+
+                Ok(t_in)
+            }
+        }
+    }
+
+    /// Traverses a term to collect the types of given ids.
+    ///
+    /// LetRec doesn't have type info stored in it. Until we change the MIR to track that information explicitly, we have to rebuild it from looking at the term.
+    fn collect_recursive_variable_types(
+        &self,
+        ids: &[LocalId],
+        ctx: &mut Ctx,
+    ) -> Result<(), TypeError> {
+        match self {
+            MirRelationExpr::Get {
+                id: Id::Local(id),
+                typ,
+            } => {
+                if !ids.contains(id) {
+                    return Ok(());
+                }
+
+                let id = Id::Local(id.clone());
+                if let Some(ctx_typ) = ctx.get_mut(&id) {
+                    for (base_col, input_col) in ctx_typ.iter_mut().zip_eq(typ.column_types.iter())
+                    {
+                        *base_col =
+                            base_col
+                                .union(&input_col)
+                                .map_err(|e| TypeError::MismatchColumn {
+                                    source: self,
+                                    got: input_col.clone(),
+                                    expected: base_col.clone(),
+                                    message: format!(
+                                        "couldn't compute union of collected column types: {:?}",
+                                        e
+                                    ),
+                                })?;
+                    }
+                } else {
+                    ctx.insert(id, typ.column_types.clone());
+                }
+            }
+            MirRelationExpr::Get {
+                id: Id::Global(..), ..
+            }
+            | MirRelationExpr::Constant { .. } => (),
+            MirRelationExpr::Let { id, value, body } => {
+                value.collect_recursive_variable_types(ids, ctx)?;
+
+                // we've shadowed the id
+                if ids.contains(id) {
+                    warn!("id {} shadowed in {:#?}", id, self);
```

Comment:
Style nit:
```suggestion
                    warn!("id {id} shadowed in {self:#?}");
```

Reply:
:exploding_head: did not know about this, thank you!